src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1450,4 +1450,76 @@ impl Cage {
             }
         }
     }
+    
+    //------------------GETDENTS SYSCALL------------------
+
+    pub fn getdents_syscall(&self, fd: i32, bufsize: usize, vec: &mut Vec<(ClippedDirent, Vec<u8>)>) -> i32 {
+        
+        // make sure bufsize is at least greater than size of a ClippedDirent struct
+        let dirent_size = std::mem::size_of::<ClippedDirent>();
+        if bufsize <= dirent_size {
+            return syscall_error(Errno::EINVAL, "getdents", "Buffer size too small");
+        }
+        
+        let fdtable = self.filedescriptortable.read().unwrap();
+
+        if let Some(wrappedfd) = fdtable.get(&fd) { // check if fd is valid
+            let mut filedesc_enum = wrappedfd.write().unwrap();
+            
+            match &mut *filedesc_enum {
+                // only proceed when fd represents a file
+                File(ref mut normalfile_filedesc_obj) => {
+                    let metadata = FS_METADATA.read().unwrap();
+                    let inodeobj = metadata.inodetable.get(&normalfile_filedesc_obj.inode).unwrap();
+
+                    match inodeobj {
+                        // only proceed when inode is a dir
+                        Inode::Dir(dir_inode_obj) => {
+                            let position = normalfile_filedesc_obj.position;
+                            let mut bufcount = 0;
+                            let mut curr_size = 0;
+                            let mut curr_pos = 0;
+                            let mut vec_idx = 0;
+                            let mut count = 0;
+
+                            for (filename, inode) in dir_inode_obj.filename_to_inode_dict.clone().into_iter() {
```

Comment:
You don't actually insert anything into vec!
What I'd do is first construct and fill in the name u8 vector, then update bufcount, and construct the clipped dirent using a struct instantiator (i.e. `ClippedDirent { d_ino: inode, d_off: bufcount, d_reclen: curr_size}` of course with the requisite casts), and then insert the tuple containing these into vec like `vec.push((curdirent, namevec))`

Reply:
Also, I think the d_off in the last dirent in the syscall needs to have some value placed into it, what value exactly that is, I'm not sure, some tests would help you figure it out. The man page test uses 4096, but I'm not sure whether that depends on the number of directory entries or not. If you'd like some help testing I'd be happy to help you.

Reply:
I did some research online and what people usually do is just leave d_off in the last dirent as zero, so I didn't make change on this one. Otherwise, I think everything mentioned has been fixed.